### ISSUE TITLE
Java setter method invocation fails in an OSGi environment after a bundle redeploy

### ISSUE THREAD

**Issue by matteozanon** [mrj0/jep#17]
Hi, I'm testing jep in an OSGi environment.
In a bundle of mine I use Jep to invoke a python function that returns at the end of computation a simple java bean tree.
The Java bean tree is instantiated from mirrorred Python objects that are used during the computation and which get converted recursively before returning from Jep.  

This is how I dynamically instantiate a Java bean from the corresponding Python object:

    def load_java_obj(self, j_class_name):
        package_and_class = j_class_name.rsplit(".", 1)
        class_name = package_and_class[1]
        java_module = importlib.import_module(package_and_class[0])
        java_type = getattr(java_module, class_name)
        return java_type()

and here is how I call the setter method:

    outer_obj = load_java_obj("org.example.MyBean")
    nested_obj = load_java_obj("org.example.MyNestedBean")
    # populate nested_obj properties
    setFunct = getattr(outer_obj, "setter_name")
    setFunct(nested_obj)

I am experiencing the following issue: the above code starts to fail if I redeploy the bundle declaring the Java classes. In the code when the redeploy takes place I create a new Jep instance and I pass the correct class loader, but after the redeploy any call to the setter methods that accept a java bean from Python leads to the following error message: Incorrect object type at 1

I am pretty sure it's not me passing in the wrong class loader to the Jep instance, in fact if I change the setters to accept a java.lang.Object and cast to the correct type within the setter body, the assignement succeeds, meaning that the objects created by Python (Jep) are instances of the classes defined by the appropriate class loader, otherwise I would experience a ClassCastException.

The error message is raised by the native C code of Jep which makes an assignement compatibility check for method arguments before calling a java method from python. 
```
            // check object itself is assignable to that type.
            if(!(*env)->IsAssignableFrom(env,
                                         ((PyJobject_Object *) param)->clazz,
                                         paramType)) {
                PyErr_Format(PyExc_TypeError,
                             "Incorrect object type at %i.",
                             pos + 1);
                return ret;
            }
```
This is why, changing the setter method parameter from a specific type to java.lang.Object makes the invocation successfull in spite of the bundle redeploy.

In practice, it appears that after a bundle redeploy the Jep C code is not correctly recognizing the java bean class anymore, as if the Java class had been defined by a different class loader. But as already written, I am passing in the correct class loader to the new Jep instance. My worry is that Jep is somehow leaking class loaders and re-using the previously passed class loader which is of course invalid after the bundle redeploy. 

As a work around to this problem, instead of calling the Java setter on the bean from Python/Jep, I am now using a more generic helper method in Java which I inwoke from Jep/Python passing in the target object, the setter name, and the setter argument. In this way, by using reflection in Java for invoking the setter, I am bypassing the assignment compatibility check mentioned before.
Alternatively, I could simply comment out such check from the C code and rebuild.

Any idea or recommendation? Any test I can do to help you diagnose the problem?

Thanks in advance.
Matteo.

**Comment by ndjensen** [comment]
Just to make sure I understand, after the bundle is redeployed, the method signature in question did not change, right?  But then the method invocation stops working cause it doesn't think the object matches up.

I'm not sure why it would fail at that point, but I suspect it's related to the caching I added in 3.3.  For performance optimization I had the C library retain a map/dictionary of java fully qualified names to Method objects.  Nothing is ever really clearing from that cache, so if the bundle is reloaded it would potentially have the wrong method reference.  See https://github.com/mrj0/jep/blob/dev_3.4/src/jep/pyjobject.c#L307

Given your scenario, I think we need to rethink that performance optimization.  It's probably also not secure in the sense that that cache in memory can have classes loaded from different classloaders.  The simplest fix is probably to tie the cache of classname to methods to the Jep instance/interpreter in question, instead of it being global across interpreters.  That seems relatively minor so I could probably get that working in dev_3.4 branch.

We do have plans to rework all that caching and determination of attributes/methods for efficiency and cleanliness, but that's a rather large task and not coming anytime soon.

**Comment by matteozanon** [comment]
Your understanding is correct.

The idea of coupling the cache life-cycle to that of the Jep instance looks good to me, although that will imply a larger memory footprint.

Alternatively, being my use case probably not that common  (in practice, I am redefining classes on a live application, which is not that frequent in JavaSE or JavaEE), a simple solution could consist in providing a static method to let client code invalidate the global cache. Then, I could just call such method when the bundle is redeployed.

**Comment by ndjensen** [comment]
It'd still be faster and about the same footprint as what was there prior to 3.3, since back then every PyJobject that was instantiated had to do reflection lookups and make new PyJmethods.  The concept of a global cache seems a bit wrong since it's potentially not respecting the ClassLoader and an OSGi environment could theoretically have two classes of the same name and different versions co-exist in different interpreters.

**Comment by ndjensen** [comment]
Ok I put a fix up on the dev_3.4 branch.  The extra memory usage and slightly more reflection calls should be negligible.  I need to do a bit more testing of it in multi-threaded environments to verify I didn't mess up the memory management, but in theory it should work and resolve the issue.  Can you try rebuilding the library and see if it does indeed resolve your issue?

**Comment by matteozanon** [comment]
Now It works like a charm. Thank you for being so helpful.

**Comment by ndjensen** [comment]
Fixed in 3.4.

**Comment by matteozanon** [comment]
I have this issue again after updating jep project to the last commit. Before I was aligned to this commit  https://github.com/mrj0/jep/tree/778b516930bdc11ecbb7751d560d373964009639 
If I rollback to that commit then everything works again.

**Comment by ndjensen** [comment]
That is unfortunate, I introduced a memory leak of references on that commit.  I just pushed up a one line fix that should take care of the memory more safely.  Can you test the latest on branch dev_3.4?  I'm hoping that that will fix it....

**Comment by matteozanon** [comment]
I try the last version and now works everything again. Thank you very much!

**Comment by ndjensen** [comment]
Great!  Thanks for testing.  Closing as fixed in 3.4.
